ui/src/components/tag/AddTag.js

```diff
@@ -0,0 +1,124 @@
+import AppUtils from '../utils/AppUtils';
+
+export default class AddTag extends React.Component {
+    constructor(props) {
+        super(props);
+        this.api = this.props.api;
+
+        if (props.editedTagKey && props.editedTagValues) {
+            this.state = {
+                editMode: true,
+                showModal: !!this.props.showAddTag,
+                tagName: props.editedTagKey,
+                tagValues: props.editedTagValues,
+                newTagValue: '',
+            };
+        } else {
+            this.state = {
+                showModal: !!this.props.showAddTag,
+                tagName: '',
+                tagValues: [],
+                newTagValue: '',
+            };
+        }
+    }
+
+    onSubmit() {
+        if (!this.state.tagName || this.state.tagName === '') {
+            this.setState({
+                errorMessage: 'Tag name is required.',
+            });
+            return;
+        }
+        if (
+            !this.props.editMode &&
+            this.props.validateTagExist(this.state.tagName)
+        ) {
+            this.setState({
+                errorMessage: 'Tag already exist.',
+            });
+            return;
+        }
+
+        if (
+            !this.state.newTagValue &&
+            (!this.state.tagValues || this.state.tagValues.length === 0)
+        ) {
+            this.setState({
+                errorMessage: 'Tag value is required.',
+            });
+            return;
+        }
+        let notAddedTag = this.state.newTagValue.trim();
+        if (notAddedTag) {
+            let tagValues = AppUtils.deepClone(this.state.tagValues);
+            tagValues.push(notAddedTag);
+            this.setState(
+                {
+                    tagValues: tagValues,
+                },
+                () =>
+                    this.props.addNewTag(
+                        this.state.tagName,
+                        this.state.tagValues
+                    )
+            );
+        } else {
+            this.props.addNewTag(this.state.tagName, this.state.tagValues);
+        }
+    }
+
+    onUpdate(tagKey, newTagValue, tagValues) {
+        this.setState({
+            tagName: tagKey,
+            newTagValue: newTagValue,
+            tagValues: tagValues,
+        });
+    }
+
+    render() {
+        const title = this.props.editMode
+            ? 'Edit ' + this.state.tagName + ' Tag'
+            : 'Add Tag to ' + this.props.resource;
+        return (
+            <AddModal
+                isOpen={this.state.showModal}
+                cancel={this.props.onCancel}
+                submit={() => this.onSubmit()}
+                title={title}
+                errorMessage={
+                    this.props.errorMessage || this.state.errorMessage
+                }
+                sections={
+                    <AddTagForm
+                        api={this.api}
+                        onUpdate={(tagKey, newTagValue, tagValues) =>
```

Comment:
we should bind this.onUpdate in the constructor and pass it here

Reply:
trying to avoid creating an anonymous function on every render

Reply:
Actually, I've changed all new tag components not to use bind.
1. The bind syntax is unreadable.
2. Performance is better (at least not worse than bind).